### PR TITLE
JIT: enable cloning based on loop invariant type tests

### PR DESCRIPTION
**Pull request by AndyAyersMS**
Such as those added by GDV.

The JIT will now clone just for type tests, or just for array bounds, or for
a mixture of the two. The JIT will still produce just one fast and one slow loop.
If there are a mixture of array bounds and type test conditions, all conditions
must pass for control to reach the fast loop.

Unlike array bounds checks, type test failures are not intrinsically rare,
so there is some profitability screening to ensure that a failed type test does
not force execution to run the slow version "too often". The type test must
execute frequently within the loop, and be heavily biased towards success.

This is work towards resolving #65206.

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
@BruceForstall  PTAL
cc @dotnet/jit-contrib 

This doesn't kick in as often as I would have expected, but the mechanical parts are there. I need to do more analysis to decide if the limitations are in loop recognition, invariant analysis, profitability, or something else.

File: src/tests/JIT/opt/GuardedDevirtualization/enumerablecloning.cs [hunk 1]
```diff
@@ -0,0 +1,48 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
Comment (BruceForstall) [inline src/tests/JIT/opt/GuardedDevirtualization/enumerablecloning.cs:1]: This is one of those tests where I wish we had the ability to verify that a particular optimization kicked in.

File: src/coreclr/jit/loopcloning.h [hunk 2]
```diff
@@ -316,6 +316,20 @@ struct LcJaggedArrayOptInfo : public LcOptInfo
...
 };
 
+// Optimization info for a type test
+//
+struct LcTypeTestOptInfo : public LcOptInfo
+{
+    // local whose method table is tested
+    unsigned lclNum;
+    // handle being tested for
+    ssize_t clsHnd;
```
Comment (BruceForstall) [inline src/coreclr/jit/loopcloning.h:326]: Why is a handle `ssize_t` type instead of a handle type?

File: src/coreclr/jit/loopcloning.h [hunk 3]
```diff
@@ -421,10 +435,12 @@ struct LC_Ident
         Var,
         ArrLen,
         Null,
+        ClassHandle,
+        Indir,
     };
 
     LC_Array  arrLen;   // The LC_Array if the type is "ArrLen"
-    unsigned  constant; // The constant value if this node is of type "Const", or the lcl num if "Var"
+    ssize_t   constant; // The constant value if this node is of type "Const", or the lcl num if "Var"
```
Comment (BruceForstall) [inline src/coreclr/jit/loopcloning.h:443]: Maybe this should be a `union`? Seems like that would eliminate a lot of casts (and make this struct smaller).

**Review by BruceForstall — COMMENTED** [review]
Looks really good. A few minor comments.

File: src/coreclr/jit/loopcloning.h (hunk 3 above)
Comment (AndyAyersMS) [reply src/coreclr/jit/loopcloning.h:443]: Yeah, I thought about doing that... makes sense.

**Comment by AndyAyersMS** [comment]
Some data (now a bit stale) about the opportunities in ASP.NET SPMI
* 323 type tests in loops
* 78 loop invariant type tests in loops
* 14 profitable loop invariant type tests in loops

All of these seem low, so I think we're losing opportunities at each stage. Invariant analysis was recently improved so I need to regather this data.

**Comment by AndyAyersMS** [comment]
SPMI legs seems to be failing to find things to download and so failing.
```
No Azure Storage MCH files to download from 5868685e-b877-4ef5-83f0-73d601e50013/windows/arm64/
```

**Comment by jakobbotsch** [comment]
That's because #70180 changed JIT-EE GUID so we need a new collection, which is running right now. Sorry about that, I'll keep in mind to merge JIT-EE GUID changes at end-of-day in the future.

**Comment by AndyAyersMS** [comment]
No worries, just was puzzled for a bit.

**Comment by AndyAyersMS** [comment]
Going to spiff up the ASP.NET collection for the new GUID and then push some changes based on Bruce's feedback.

**Review by BruceForstall — APPROVED** [review]
LGTM

**Comment by AndyAyersMS** [comment]
Some perf cases that need investigation

[IEnumerable](https://pvscmdupload.blob.core.windows.net/reports/allTestHistory%2frefs%2fheads%2fmain_x64_Windows%2010.0.18362_PGOType%3dfullpgo%2fSystem.Collections.IterateForEach(Int32).IEnumerable(Size%3a%20512).html)

newplot - 2022-06-18T090143 026

(non-pgo result, for reference)


newplot - 2022-06-18T090608 558

**Pull request merged**